### PR TITLE
refactor alert_user.js + add option to have messages fade out after a fixed interval

### PR DESCRIPTION
**Pull request by snopoke**
## Summary
This adds an option to the `alert_user` JS function that will cause messages to automatically be removed from the UI after 5s. This can be useful for UI's that make ajax requests and display messages.

Peek 2021-07-28 10-24

@biyeun @orangejenny any thoughts on this? One thing I don't like is the we have these set up in the UI so that when they are removed the whole UI moves up. It would great if they could 'float' on top or something.

If we don't want to keep the 'fadeOut' option I'll remove that commit but I think the refactor is still worth keeping. Also this fixes the fade in which was broken.

Review by commit.

## Safety Assurance

- [x] Risk label is set correctly
- [x] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Safety story
This changes a small part of the HQ UI that's related to displaying alerts from JS. I have tested this locally.

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### CODE REVIEW

File: corehq/apps/hqwebapp/static/hqwebapp/js/alert_user.js [hunk 1]
```diff
@@ -21,53 +21,61 @@ function (
...
-        var tags = "alert-" + emphasis;
-        if (!append || message_alerts().length === 0) {
-            message_alerts.push(message_alert(message, tags));
-        } else {
-            var alert = message_alerts()[0];
-            alert.message(alert.message() + "<br>" + message);
-            if (!alert.alert_class().includes(tags)) {
-                alert.alert_class(alert.alert_class() + ' ' + tags);
+    const ViewModel = function() {
+        let self = this;
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/alert_user.js:48]: Won't this set self to the window?

File: corehq/apps/hqwebapp/static/hqwebapp/js/alert_user.js [hunk 2]
```diff
@@ -25,7 +25,7 @@ function (
         var self = {
             "message": ko.observable(message),
             "alert_class": ko.observable(
-                "alert fade in message-alert"
+                "alert fadeIn message-alert"
```
Comment (orangejenny) [inline corehq/apps/hqwebapp/static/hqwebapp/js/alert_user.js:29]: Do you know where `fadeIn` comes from? I didn't think bootstrap ever used camel case for class names, and we do use `fade in` elsewhere.

**Comment by orangejenny** [comment]
@snopoke I like it. I think fading messages have worked for us well in web apps.

> @biyeun @orangejenny any thoughts on this? One thing I don't like is the we have these set up in the UI so that when they are removed the whole UI moves up. It would great if they could 'float' on top or something.

It might not be hard to make them floating/sticky ([here's](https://github.com/dimagi/commcare-hq/pull/28418) a PR that applied that effect to web apps breadcrumbs), although I think when you're dealing with ajax requests on a long page, it's also worth considering showing success in a way other than a message at the top of the page, which probably isn't where the user is looking.

File: corehq/apps/hqwebapp/static/hqwebapp/js/alert_user.js (hunk 1 above)
Comment (snopoke) [reply corehq/apps/hqwebapp/static/hqwebapp/js/alert_user.js:48]: Ah yes, that's true.

File: corehq/apps/hqwebapp/static/hqwebapp/js/alert_user.js (hunk 2 above)
Comment (snopoke) [reply corehq/apps/hqwebapp/static/hqwebapp/js/alert_user.js:29]: It's defined in utils.less (https://github.com/dimagi/commcare-hq/blob/9757db6d9ed849e508b4c58fb3a190a687d7965d/corehq/apps/hqwebapp/static/hqwebapp/less/_hq/utils.less#L28)

I see that `fade in` is the bootstrap classes. The transition time is 0.15s so it's just very fast which is why I thought it wasn't working. I'll revert to that.

**Comment by snopoke** [comment]
> It might not be hard to make them floating/sticky ([here's](https://github.com/dimagi/commcare-hq/pull/28418) a PR that applied that effect to web apps breadcrumbs), although I think when you're dealing with ajax requests on a long page, it's also worth considering showing success in a way other than a message at the top of the page, which probably isn't where the user is looking.

I don't think I'm going to mess with that for now since that also involves messing with some core layout functions and app preview. I also don't really like how the notifications go over the content on the page. Maybe that would be OK if they didn't stick around too long.

**Review by orangejenny — APPROVED** [review]

**Pull request merged**